protocols/kad/src/record.rs

```diff
@@ -122,6 +122,21 @@ impl Hash for ProviderRecord {
     }
 }
 
+impl PartialEq for ProviderRecord {
```

Comment:
Drawing attention to this non-trivial change.

As far as I can tell `libp2p-kad` does not depend on `ProviderRecord` equality when it comes to the `expires`, or `addresses` field.

In case this looks reasonable to you, let me know in case you would like a changelog entry.

Reply:
Is this change connected in any way to clippy suggestions? I don't remember exactly, but I do think the derived `PartialEq` was deliberate at some point, but maybe the code changed afterwards. It is certainly important to know and understand these equality semantics. E.g. if a `RecordStore` or any code path discards records given because it thinks it already has that record based on a comparison with this new `PartialEq`, that would not go well, since these records are regularly re-published on purpose, and if a new record has either new addresses or a new expiry, it should be stored and replace the existing one. So we need to be very clear on when two provider records are considered equal. I'm not sure they should be considered equal if they have different addresses or a different expiry, but if you think that makes more sense, I'm fine with it, but it may be worthwhile documenting it more clearly in that case.

Reply:
Actually, this change to equality may be a real problem, because the `MemoryStore` has a `provided: HashSet<ProviderRecord>`.

Reply:
Or, looking at the `MemoryStore` implementation, it may even be a _fix_. :eyes:.

Reply:
My preliminary conclusion after reviewing that code is that it makes no difference, but the new equality semantics would be nevertheless be preferable, as I don't think that in principle `MemoryStore::provided` should ever contain two records for the same key and provider but with different expiry or addresses. What is your verdict after consideration of my comments?

Reply:
> My preliminary conclusion after reviewing that code is that it makes no difference

:+1: 

> What is your verdict after consideration of my comments?

In my eyes this newer equality semantic is more intuitive, especially given that it is on-par with the `Hash` implementation.

> but it may be worthwhile documenting it more clearly in that case.

I have pushed 1d28e21f94c84c552ef10f5011d393330bad381f. What do you think?